agent/jvm/src/main/java/org/jolokia/jvmagent/handler/JolokiaHttpHandler.java

```diff
@@ -317,6 +318,47 @@ private void sendForbidden(HttpExchange pExchange, SecurityException securityExc
     }
 
     private void sendResponse(HttpExchange pExchange, ParsedUri pParsedUri, JSONAware pJson) throws IOException {
+        boolean streaming = configuration.getAsBoolean(ConfigKey.STREAMING);
+        if (streaming) {
+            JSONStreamAware jsonStream = (JSONStreamAware)pJson;
+            sendStreamingResponse(pExchange, pParsedUri, jsonStream);
+        } else {
+            sendAllJSON(pExchange, pParsedUri, pJson);
+        }
+    }
+
+    private void sendStreamingResponse(HttpExchange pExchange, ParsedUri pParsedUri, JSONStreamAware pJson) throws IOException {
+        ChunkedWriter writer = null;
+        try {
+            Headers headers = pExchange.getResponseHeaders();
+            if (pJson != null) {
+                headers.set("Content-Type", getMimeType(pParsedUri) + "; charset=utf-8");
+                String callback = pParsedUri.getParameter(ConfigKey.CALLBACK.getKeyValue());
+                pExchange.sendResponseHeaders(200, 0);
+                writer = new ChunkedWriter(pExchange.getResponseBody(), "UTF-8");
```

Comment:
@gnufied one question: Why actually it is required to use a `ChunkedWriter` here ? Or stated otherwise: What would be the drawback to use the response body outputstream directly when streaming out the response ?

I'm still looking for a quick solution to get rid of the IP issue we have (and which has been escalate quite a bit), so if don't find a quick solution I'm going to remove ChunkedWriter altogether this weekend.

Reply:
@gnufied I mean, replacing `ChunkedWriter` with `OutputStreamWriter`, what would we loose or what would be the issues ?

Reply:
@gnufied If I understand the code right, then the only difference between a plain `OutputStreamWriter` and `ChunkedWriter` is, then the latter does always "full" flush on `flush()` whereas the `OutputStreamWriter` does this during `close()`. 

Why is this required ?

Reply:
There are other minor differences - `ChunkedWriter` writes a empty byte on flush , the OutputStreamWriter doesn't - https://github.com/rhuss/jolokia/blob/master/agent/core/src/main/java/org/jolokia/util/ChunkedWriter.java#L117

Reply:
So partly this is coming back to me(sorry I forgot bunch of this code) - the main reason we had to reimplement chunkedWriter from scratch is because of empty byte handling on flush. 

StreamEncoder has `OutputStream` property marked as private  `OutputStreamWriter` delegates to `StreamEncoder` for majority of its functionality anyways.

So the problem boils down to - we need to write the empty byte on flush, to mark it as end of the whole message. `OutputStreamWriter` does not do that and hence its chunking implementation is  weirdly incomplete.

Now - in `StreamEncoder` because `OutputStream` property is private, a subclass can't access it to write that empty byte. Also directly calling `streamEncoder.write(EMPTY_BYTE)` will result in message getting buffered and that is not what we want, because this was supposed to be end of whole message. We want to be able to directly write the `EMPTY` byte to `OuputputStream`.

I think that was the most crucial reason. There may be something else but I don't remember it.

Reply:
@gnufied thanks a lot for the explanation. However, I don't get why you need to write an empty byte array to the stream. Whit is this needed ? 

Also, if I'm not completely wrong, this write is a no-op anyway:

```java
private static final byte[] EMPTY = {};
...
out.write(EMPTY);
```

then in `OutputStream`:

```java
  public void write(byte b[]) throws IOException {
        write(b, 0, b.length);
   }
   ....
 public void write(byte b[], int off, int len) throws IOException {
        if (b == null) {
            throw new NullPointerException();
        } else if ((off < 0) || (off > b.length) || (len < 0) ||
                   ((off + len) > b.length) || ((off + len) < 0)) {
            throw new IndexOutOfBoundsException();
        } else if (len == 0) {
             // This is where we end up when writing an empty array. So the original 
            // operation is a complete no-op. What did I miss ?
            return;
        }
        for (int i = 0 ; i < len ; i++) {
            write(b[off + i]);
        }
    } 
```

Reply:
I need to double check but that final EMPTY byte array causes "0\r\n\r\n" to be written to the message and that is essential for chunking to be finished.